src/shared/ui/prompter.ts

```diff
@@ -65,6 +70,108 @@ export abstract class Prompter<T> {
     }
 
+    /** Sets a 'step estimator' function used to predict how many steps are remaining in a given flow */
+    public abstract setStepEstimator(estimator: StepEstimator<T>): void
     protected abstract promptUser(): Promise<PromptResult<T>>
     public abstract setSteps(current: number, total: number): void
 }
+
+export interface CachedPrompter<T, TState extends Record<string, unknown>> {
+    (state: StateWithCache<TState, T>): Prompter<T>
+}
+
+type Cacheable = (...args: (string | number | boolean | undefined)[]) => NonNullable<any>
+
+/**
+ * A wrapped function that can cache both synchronous and asynchronous return types.
+ *
+ * Pending promises are returned as-is, while resolved promises are 'unboxed' into their promised type.
+ */
+export interface CachedFunction<F extends (...args: any) => any> {
+    (...args: Parameters<F>): ReturnType<F> extends Promise<infer Inner> ? ReturnType<F> | Inner : ReturnType<F>
+    clearCache(): void
+    supplantLast(result: ReturnType<F> extends Promise<infer Inner> ? Inner : ReturnType<F>): void
+}
+
+// TODO: this currently just caches primitive arguments which is not very flexible
+// there is a 'cache-object' library that could make sense here. we could just cache the entire
+// argument list itself and not worry about types
+function createCachedFunction<F extends Cacheable>(
+    loader: F,
+    cache: { [key: string]: ReturnType<F> } = {},
+    keys: Set<string> = new Set()
+): CachedFunction<F> {
+    const wrapped = (...args: Parameters<F>) => {
+        const key =
+            args
+                .map(arg => (arg ?? '').toString())
+                .map(arg => (arg === '' ? arg : `${arg}${arg.length}`))
+                .join() + '0' // Cannot index with an empty string
+
+        if (cache[key] !== undefined) {
+            return cache[key]
+        }
+
+        const resolved = loader(...args)
+        cache[key] = resolved as any
+        keys.add(key)
+
+        if (resolved instanceof Promise) {
+            return resolved.then(result => {
+                cache[key] = result
+                return result
+            })
+        }
+
+        return resolved
+    }
+
+    const clearCache = () => {
+        keys.forEach(key => delete cache[key])
+        keys.clear()
+    }
+
+    const supplantLast = (result: ReturnType<F>) => {
+        if (keys.size === 0) {
+            throw new Error('Cannot evict an empty cache')
+        }
+        cache[[...keys.values()].pop()!] = result
+    }
+
+    return Object.assign(wrapped, { clearCache, supplantLast })
+}
+
+// Rebinds the Function function body to instead call the extended class method
+const EXECUTE_CALLEE = 'return arguments.callee.call.apply(arguments.callee, arguments)'
+/**
+ * Convenience class that lightly wraps the creation of a Prompter with the loading of whatever resources
+ * are required. The 'load' call is wrapped with a cache for automatic caching of its results.
+ */
+export abstract class CachedPrompter<
```

Comment:
I'm probably going to get rid of this class. Seemed like a good idea at first but involves too much indirection.

Reply:
Does that include CachedFunction? The Cached stuff may warrant a separate PR.

Is CachedFunction intended to act like memoize ([python](https://docs.python.org/3/library/functools.html#functools.lru_cache)) ?

Reply:
> Does that include CachedFunction? The Cached stuff may warrant a separate PR.

I'll clean it up a bit and turn it into its own PR. It's in a somewhat weird state since I was wanting it to handle sync and async functions, though I think it'd be more reasonable to separate them.

> Is CachedFunction intended to act like memoize (python) ? Would be helpful to put that in its docstring.

Yeah essentially like the `lru_cache` decorator but without all the fancy options. There's also a twist to it since it will cache Promises too. Future calls return the Promise while it is still pending, but after it resolves it returns the final value of the Promise. This was done to prevent weird UI jitters for things that were promises but immediately resolved.